File: Project_code/PDFS.cpp
#include <iostream>
#include <vector>
#include <list>
#include <algorithm>

using namespace std;

// Structure to represent an edge in the tree
struct Edge
{
    int to;
    double weight;
};

// Structure to represent a tree
class Tree
{
private:
    int n;                    // Number of vertices
    int root;                 // Root of the tree
    vector<vector<Edge>> adj; // Adjacency list
    vector<bool> visited;     // To track visited nodes during DFS

    // Helper function for DFS traversal
    void dfs(int v, vector<int> &rdfs)
    {
        visited[v] = true;
        rdfs.push_back(v);

        for (const Edge &e : adj[v])
        {
            if (!visited[e.to])
            {
                dfs(e.to, rdfs);
                rdfs.push_back(v); // Return to parent after exploring subtree
            }
        }
    }

public:
    Tree(int n, int root) : n(n), root(root)
    {
        adj.resize(n);
        visited.resize(n, false);
    }

    // Add an undirected edge to the tree
    void addEdge(int u, int v, double weight)
    {
        adj[u].push_back({v, weight});
        adj[v].push_back({u, weight});
    }

    // Generate the RDFS traversal
    vector<int> generateRDFS()
    {
        vector<int> rdfs;
        fill(visited.begin(), visited.end(), false);
        dfs(root, rdfs);
        return rdfs;
    }

    // Calculate distance between two vertices
    double distance(int u, int v)
    {
        if (u == v)
            return 0;

        vector<bool> vis(n, false);
        vector<double> dist(n, -1);
        list<int> queue;

        vis[u] = true;
        dist[u] = 0;
        queue.push_back(u);

        while (!queue.empty())
        {
            int s = queue.front();
            queue.pop_front();

            for (const Edge &e : adj[s])
            {
                if (!vis[e.to])
                {
                    vis[e.to] = true;
                    dist[e.to] = dist[s] + e.weight;

                    if (e.to == v)
                        return dist[e.to];

                    queue.push_back(e.to);
                }
            }
        }

        return -1; // Should not reach here in a connected tree
    }

    // Calculate the length of a path
    double pathLength(const vector<int> &path)
    {
        double length = 0;
        for (size_t i = 1; i < path.size(); i++)
        {
            // Find the edge weight between path[i-1] and path[i]
            for (const Edge &e : adj[path[i - 1]])
            {
                if (e.to == path[i])
                {
                    length += e.weight;
                    break;
                }
            }
        }
        return length;
    }

    // Find the path from u to v
    vector<int> findPath(int u, int v)
    {
        if (u == v)
            return {u};

        vector<int> parent(n, -1);
        vector<bool> vis(n, false);
        list<int> queue;

        vis[u] = true;
        queue.push_back(u);

        while (!queue.empty())
        {
            int s = queue.front();
            queue.pop_front();

            for (const Edge &e : adj[s])
            {
                if (!vis[e.to])
                {
                    vis[e.to] = true;
                    parent[e.to] = s;

                    if (e.to == v)
                    {
                        // Construct the path
                        vector<int> path;
                        int current = v;
                        while (current != -1)
                        {
                            path.push_back(current);
                            current = parent[current];
                        }
                        reverse(path.begin(), path.end());
                        return path;
                    }

                    queue.push_back(e.to);
                }
            }
        }

        return {}; // Should not reach here in a connected tree
    }
};

// Structure to represent a route
struct Route
{
    vector<int> vertices;
    double length;
};

// Calculate PDFS strategy
vector<Route> calculatePDFS(Tree &tree, double B)
{
    vector<Route> routes;
    vector<int> rdfs = tree.generateRDFS();
    int root = rdfs[0];

    int j_prev = 0; // Start from the root

    while (j_prev < rdfs.size())
    {
        Route route;
        int j_current = j_prev;

        // Find the farthest point we can visit with energy budget B
        for (int p = j_prev + 1; p < rdfs.size(); p++)
        {
            // Calculate the tentative route: root -> j_prev -> p -> root
            vector<int> path_to_j_prev = tree.findPath(root, rdfs[j_prev]);

            // The progress part is simply the subarray of rdfs from j_prev to p
            vector<int> progress_part;
            for (int idx = j_prev; idx <= p; idx++)
            {
                progress_part.push_back(rdfs[idx]);
            }

            vector<int> path_to_root = tree.findPath(rdfs[p], root);

            // Calculate total length
            double total_length = tree.pathLength(path_to_j_prev) +
                                  tree.pathLength(progress_part) +
                                  tree.pathLength(path_to_root);

            if (total_length <= B)
            {
                j_current = p;
            }
            else
            {
                break; // Exceeds budget, stop here
            }
        }

        // Construct the route
        vector<int> path_to_j_prev = tree.findPath(root, rdfs[j_prev]);

        // The progress part
        vector<int> progress_part;
        for (int idx = j_prev; idx <= j_current; idx++)
        {
            progress_part.push_back(rdfs[idx]);
        }

        vector<int> path_to_root = tree.findPath(rdfs[j_current], root);

        // Combine the parts to form the complete route
        route.vertices = path_to_j_prev;
        // Remove the last element to avoid duplication
        if (!path_to_j_prev.empty() && !progress_part.empty() &&
            path_to_j_prev.back() == progress_part.front())
        {
            route.vertices.pop_back();
        }
        route.vertices.insert(route.vertices.end(), progress_part.begin(), progress_part.end());

        // Remove the last element to avoid duplication
        if (!progress_part.empty() && !path_to_root.empty() &&
            progress_part.back() == path_to_root.front())
        {
            route.vertices.pop_back();
        }
        route.vertices.insert(route.vertices.end(), path_to_root.begin(), path_to_root.end());

        route.length = tree.pathLength(route.vertices);
        routes.push_back(route);

        // Move to the next starting point
        j_prev = j_current + 1;

        // If we've reached the end of rdfs, we're done
        if (j_current == rdfs.size() - 1)
        {
            break;
        }
    }

    return routes;
}

// Print a route
void printRoute(const Route &route)
{
    cout << "Route: ";
    for (int vertex : route.vertices)
    {
        cout << vertex << " ";
    }
    cout << "Length: " << route.length << endl;
}

void inputTree(Tree &tree, int n)
{
    for (int i = 0; i < n - 1; i++)
    {
        int u, v;
        double weight;
        cout << "Enter edge (u, v) and weight: ";
        cin >> u >> v >> weight;
        tree.addEdge(u, v, weight);
    }
}

int main()
{
    int n; // Number of vertices
    int root;
    double B; // Energy budget
    cout << "Enter the number of vertices in the tree: ";
    cin >> n;
    cout << "Enter the energy budget B: ";
    cin >> B;
    cout << "Enter the root of the tree: ";
    cin >> root;
    if (n <= 0 || B <= 0)
    {
        cerr << "Invalid input. Number of vertices and energy budget must be positive." << endl;
        return 1;
    }
    // Create a tree similar to the one in Fig.1 of the problem statement
    Tree tree(n, root);
    cout << "Input the edges of the tree:" << endl;
    inputTree(tree, n);

    vector<Route> pdfs = calculatePDFS(tree, B);

    // Print the routes
    cout << "PDFS Strategy with B = " << B << ":" << endl;
    for (size_t i = 0; i < pdfs.size(); i++)
    {
        cout << "R" << i + 1 << ": ";
        printRoute(pdfs[i]);
    }

    cout << "Total number of routes: " << pdfs.size() << endl;
}

File: Project_code/PDFS.py
from dataclasses import dataclass
from typing import List, Dict
from collections import deque


@dataclass
class Edge:
    to: int
    weight: float


class Tree:
    def __init__(self, n: int, root: int):
        self.n = n
        self.root = root
        self.adj = [[] for _ in range(n)]
        self.visited = [False] * n

    def add_edge(self, u: int, v: int, weight: float):
        self.adj[u].append(Edge(v, weight))
        self.adj[v].append(Edge(u, weight))

    def _dfs(self, v: int, rdfs: List[int]):
        self.visited[v] = True
        rdfs.append(v)

        for e in self.adj[v]:
            if not self.visited[e.to]:
                self._dfs(e.to, rdfs)
                rdfs.append(v)  # Return to parent after exploring subtree

    def generate_rdfs(self) -> List[int]:
        self.visited = [False] * self.n
        rdfs = []
        self._dfs(self.root, rdfs)
        return rdfs

    def distance(self, u: int, v: int) -> float:
        if u == v:
            return 0

        vis = [False] * self.n
        dist = [-1] * self.n
        queue = deque([u])
        vis[u] = True
        dist[u] = 0

        while queue:
            s = queue.popleft()
            for e in self.adj[s]:
                if not vis[e.to]:
                    vis[e.to] = True
                    dist[e.to] = dist[s] + e.weight
                    if e.to == v:
                        return dist[e.to]
                    queue.append(e.to)
        return -1

    def path_length(self, path: List[int]) -> float:
        length = 0
        for i in range(1, len(path)):
            for e in self.adj[path[i - 1]]:
                if e.to == path[i]:
                    length += e.weight
                    break
        return length

    def find_path(self, u: int, v: int) -> List[int]:
        if u == v:
            return [u]

        parent = [-1] * self.n
        vis = [False] * self.n
        queue = deque([u])
        vis[u] = True

        while queue:
            s = queue.popleft()
            for e in self.adj[s]:
                if not vis[e.to]:
                    vis[e.to] = True
                    parent[e.to] = s
                    if e.to == v:
                        path = []
                        current = v
                        while current != -1:
                            path.append(current)
                            current = parent[current]
                        return path[::-1]
                    queue.append(e.to)
        return []


@dataclass
class Route:
    vertices: List[int]
    length: float


def calculate_pdfs(tree: Tree, B: float) -> List[Route]:
    routes = []
    rdfs = tree.generate_rdfs()
    root = rdfs[0]
    j_prev = 0

    while j_prev < len(rdfs):
        route = Route(vertices=[], length=0)
        j_current = j_prev

        # Find farthest point we can visit
        for p in range(j_prev + 1, len(rdfs)):
            path_to_j_prev = tree.find_path(root, rdfs[j_prev])
            progress_part = rdfs[j_prev : p + 1]
            path_to_root = tree.find_path(rdfs[p], root)

            total_length = (
                tree.path_length(path_to_j_prev)
                + tree.path_length(progress_part)
                + tree.path_length(path_to_root)
            )

            if total_length <= B:
                j_current = p
            else:
                break

        # Construct route
        path_to_j_prev = tree.find_path(root, rdfs[j_prev])
        progress_part = rdfs[j_prev : j_current + 1]
        path_to_root = tree.find_path(rdfs[j_current], root)

        # Combine parts
        route.vertices = path_to_j_prev
        if path_to_j_prev and progress_part and path_to_j_prev[-1] == progress_part[0]:
            route.vertices.pop()
        route.vertices.extend(progress_part)

        if progress_part and path_to_root and progress_part[-1] == path_to_root[0]:
            route.vertices.pop()
        route.vertices.extend(path_to_root)

        route.length = tree.path_length(route.vertices)
        routes.append(route)

        j_prev = j_current + 1
        if j_current == len(rdfs) - 1:
            break

    return routes


def print_route(route: Route, index: int = None):
    route_str = "R{}: ".format(index) if index is not None else "Route: "
    vertices_str = " ".join(str(v) for v in route.vertices)
    print(f"{route_str}{vertices_str} Length: {route.length}")


def input_tree(tree: Tree, n: int):
    for i in range(n - 1):
        print("Enter edge (u, v) and weight: ", end="")
        u, v, weight = map(float, input().split())
        tree.add_edge(int(u), int(v), weight)


def main():
    print("Enter the number of vertices in the tree: ", end="")
    n = int(input())
    print("Enter the energy budget B: ", end="")
    B = float(input())
    print("Enter the root of the tree: ", end="")
    root = int(input())

    if n <= 0 or B <= 0:
        print("Invalid input. Number of vertices and energy budget must be positive.")
        return

    tree = Tree(n, root)
    print("Input the edges of the tree:")
    input_tree(tree, n)

    pdfs = calculate_pdfs(tree, B)

    print(f"\nPDFS Strategy with B = {B}:")
    for i, route in enumerate(pdfs):
        print_route(route, i + 1)

    print(f"Total number of routes: {len(pdfs)}")


if __name__ == "__main__":
    main()
